### PR TITLE
Refactor DefaultQueryPlanner into discrete methods

### PR DESCRIPTION
**Pull request by apmoriarty**
This is a proof of concept merge request. The goal is to eliminate boilerplate code in the DefaultQueryPlanner while also providing more options to override and extend specific operations.

### CODE REVIEW

**Review by FineAndDandy — DISMISSED** [review]

**Comment by jwomeara** [comment]
I think you're on to something good here.  You should push forward with the other visitors and add in configuration which allows us to conditionally run validity checks after each visitor is run (e.g. like validate lineage).  I see that the roots of that are already there.  keep going!

**Review by jwomeara — CHANGES_REQUESTED** [review]
I would encapsulate the stuff after the processTree call as well.  I know that it's a little bit heavier weight, but in my opinion, that's even more of a reason to pull it into a separate function.  One of the goals of this update in my mind would be to make it easier to create an extended version of this class.  Encapsulating those other blocks of code would help with that.  

Also, consider making TimedVisitorManager a standalone, package-private class in the same package as the DefaultQueryPlanner.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 1]
```diff
@@ -961,142 +859,43 @@ protected ASTJexlScript processTree(final ASTJexlScript originalQueryTree, Shard
...
+                        getTransformRules());
         
         // Find unfielded terms, and fully qualify them with an OR of all fields
         // found in the index
         // If the max term expansion is reached, then the original query tree is
         // returned.
         // If the max regex expansion is reached for a term, then it will be
         // left as a regex
-        Set<String> expansionFields = null;
+        Set<String> expansionFields = loadExpansionFields(config);
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: this should only be executed if !disabledAnyFieldLookup

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 2]
```diff
@@ -961,142 +859,43 @@ protected ASTJexlScript processTree(final ASTJexlScript originalQueryTree, Shard
...
+            timedTestForNonExistentFields(timers, queryTree, config, metadataHelper, queryModel, settings);
         }
         
         // apply the node transform rules
         // running it here before any regex or range expansions to enable potentially pushing down terms before index lookups
-        queryTree = applyNodeTransformRules(queryTree, getTransformRules(), config, metadataHelper, "Pre regex/range expansions");
+        queryTree = timedApplyNodeTransformRules(timers, "Apply Node Transform Rules - Pre Regex/Range Expansions", queryTree, config, metadataHelper,
+                        getTransformRules());
         
         stopwatch = timers.newStartedStopwatch("DefaultQueryPlanner - Fetch required dataTypes");
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: add timer encapsulation for this and other steps that aren't getting the tree updates

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 3]
```diff
@@ -1280,35 +1007,14 @@ protected ASTJexlScript processTree(final ASTJexlScript originalQueryTree, Shard
...
-                    innerStopwatch = timers.newStartedStopwatch("DefaultQueryPlanner - Expand pushing functions into exceeded value ranges");
-                    queryTree = PushFunctionsIntoExceededValueRanges.pushFunctions(queryTree, metadataHelper, config.getDatatypeFilter());
-                    if (log.isDebugEnabled()) {
-                        logQuery(queryTree, "Query after expanding pushing functions into exceeded value ranges:");
-                    }
-                    innerStopwatch.stop();
+                    queryTree = timedPushFunctions(timers, queryTree, config, metadataHelper);
                 }
                 
                 innerStopwatch = timers.newStartedStopwatch("DefaultQueryPlanner - Remove delayed predicates");
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: hit the other tree expansions below

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 4]
```diff
@@ -2272,6 +2281,93 @@ private RangeStream initializeRangeStream(ShardQueryConfiguration config, Scanne
         }
     }
     
+    /**
+     * Handles boilerplate code execution for operations like timing a visit call, logging the query tree, and eventualy validating the resulting query tree.
+     */
+    public class TimedVisitorManager {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: can this be pulled into its own class?

**Review by FineAndDandy — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: done

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: done.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: done

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 4 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: done

**Comment by apmoriarty** [comment]
Rebased to address merge conflict introduced by https://github.com/NationalSecurityAgency/datawave/pull/1244, extracted that code into a timed method call.

**Review by jwomeara — DISMISSED** [review]
Thank you for doing this!  Good work!

**Review by FineAndDandy — APPROVED** [review]

**Comment by ivakegg** [comment]
I think this is great.  My only comment is that naming each method starting with "timed" seems excessive.  Other than that this appears good.

**Review by ivakegg — APPROVED** [review]

**Pull request merged**